### PR TITLE
[5.0] Drop es5 js files

### PR DESCRIPTION
**Pull request by dgrammatiko**
Pull Request for Issue # .

### Summary of Changes

- ES5 js files are targeting browser before 2015!
- Joomla 5 (the js parts) would work on anything ES2018 or newer (that's 5 year old browsers!) 
- The 2018 was chosen as the year that all browsers supported ES Modules https://caniuse.com/es6-module
- The PR adjusts all the joomla.asset.json files
- It removes 2 dead npm packages `chosen` and `focus-visible` (never used in v4, the first was rolled with v4 for b/c the second was removed in some beta but the dependency was left behind)
- the build tools were adjusted

### Testing Instructions

Joomla still works

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [ ] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [ ] No documentation changes for manual.joomla.org needed

@nibra @HLeithner this **is** my take on what the deliverables should be but there was also a discussion prior to v4 release on the same subject: 
- https://github.com/joomla/joomla-cms/discussions/32148
- https://github.com/joomla/joomla-cms/discussions/35737

These conversations actually led to shipping type=module/nomodule on the first place but I guess now it's time to skip the es5 altogether.

### CODE REVIEW

File: build/build-modules-js/settings.json [hunk 1]
```diff
@@ -73,24 +73,12 @@
...
-              "defer": true
-            }
-          },
           {
             "name": "bootstrap.alert",
             "type": "script",
             "uri": "alert.min.js",
             "dependencies": [
-              "bootstrap.es5"
+              "core"
```
Comment (laoneo) [inline build/build-modules-js/settings.json:94]: Does this mean, BS is a dependency of core?

Comment (dgrammatiko) [reply build/build-modules-js/settings.json:94]: Yes, all scripts use Joomla.getOptions()

Comment (laoneo) [reply build/build-modules-js/settings.json:94]: Is then no bootstrap library needed anymore for the individual bootstrap scripts like bootstrap.alert?

Comment (dgrammatiko) [reply build/build-modules-js/settings.json:94]: This is all about the packaging, each component has an import of the dom.js and popover, tooltip and drop-down also have an import of the popper.js. But, for the esm version there was never a bootstrap.js, if someone wrongly use the old JHtml::_(‘bootstrap.core’) they got back all the esm files

Comment (HLeithner) [reply build/build-modules-js/settings.json:94]: how does this effect existing templates using our bootstrap?

Comment (dgrammatiko) [reply build/build-modules-js/settings.json:94]: Ok, probably my explanation wasn't any good, so let me try again.
The `HTMLHelper::_('bootstrap.framework')` is a function that adds **all** the Bootstrap JS components:
https://github.com/joomla/joomla-cms/blob/4b1c7b0c0412f3051cb6ac04ff9543bf2c1ec952/libraries/src/HTML/Helpers/Bootstrap.php#L574-L598

Each component has a dependency on core.js (due to `Joomla.getOptions()`). This is working fine (although it's really wasteful of bandwidth and decreasing perf due to the amount of JS delivered/parsed/executed) and although there was a discussion to drop the function for v4 it was left but **IT IS** deprecated...
Decision always up to the release leaders/PT, but regarding the code here I'm not changing anything other than removing the es5.

If the question is about any side effects for templates due to removing the es5 files: the answer is **no** because the es5 are not actually functional (for the browsers they're supposed to ran) due to missing DOM polyfills (eg HTMLElement.closest()), so essentially is dead weight the project distributes for zero benefit

File: build/build-modules-js/javascript/compile-to-es2017.es6.js [hunk 2]
```diff
@@ -99,7 +101,7 @@ module.exports.handleESMFile = async (file) => {
 
       return writeFile(resolve(`${newPath}.min.js`), content.code, { encoding: 'utf8', mode: 0o644 });
     })
-    .then(() => handleESMToLegacy(resolve(`${newPath}.js`)))
+    // .then(() => handleESMToLegacy(resolve(`${newPath}.js`)))
```
Comment (brianteeman) [inline build/build-modules-js/javascript/compile-to-es2017.es6.js]: should this line not be removed instead of just commented out?

Comment (dgrammatiko) [reply build/build-modules-js/javascript/compile-to-es2017.es6.js]: Good catch

**Comment by HLeithner** [comment]
@dgrammatiko could you please rebase this?

**Comment by dgrammatiko** [comment]
@HLeithner done

File: build/build-modules-js/javascript/build-bootstrap-js.es6.js [hunk 3]
```diff
@@ -47,22 +46,18 @@ const build = async () => {
             {
               targets: {
                 browsers: [
-                  '> 1%',
-                  'not ie 11',
-                  'not op_mini all',
+                  /** https://caniuse.com/es6-module */
+                  'chrome 61',
```
Comment (wilsonge) [inline build/build-modules-js/javascript/build-bootstrap-js.es6.js]: This looks weird. Aren't you just targeting 4 browsers rather than higher versions too?

File: build/build-modules-js/settings.json [hunk 4]
```diff
@@ -340,23 +328,6 @@
         "dependencies": [],
         "licenseFilename": "license"
       },
-      "focus-visible": {
```
Comment (wilsonge) [inline build/build-modules-js/settings.json]: Makes sense to remove this but we need to add in the css selector code for the replacement

**Review by wilsonge — CHANGES_REQUESTED** [review]
Obviously agree with the concept. Largely lgtm only final comment do we still not want to use rollup in most cases against the code so that we can use modern js syntax like async/await or similar if we want

File: build/build-modules-js/javascript/build-bootstrap-js.es6.js (hunk 3 above)
Comment (dgrammatiko) [reply build/build-modules-js/javascript/build-bootstrap-js.es6.js]: The way these entries work are like switches (or feature detection [it's not]) for babel to polyfill unsupported features. So, we set the Chrome to 61 means: up to 61 nothing needs polyfill, things added after version 61 need a polyfill (which babel might be able to provide through corejs, but that is not guaranteed also as a few things are not even polyfillable or belong to the browser domain, these need another strategy [eg `HTMLElement.prototype.cloaser()`]).
A practical example: [Chrome 61 doesn't support `String.prototype.replaceAll()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replaceAll#browser_compatibility) so if you edit core.esm.js and add `'Dimitris'.replaceAll('i', 'u');` and check the transpiled code you should have the few lines for the polyfill...

File: build/build-modules-js/settings.json (hunk 4 above)
Comment (dgrammatiko) [reply build/build-modules-js/settings.json]: Bootstrap 5.2 added the support for `focus-visible` but I guess the custom CSS needs to be updated (if there are any instances). That should be done in another PR. If this ever get merged we should open a couple issues (focus-visible, remove `DOMContentLoaded` events, etc)

**Comment by HLeithner** [comment]
can we split this pr please, one "simple" pr that only removes es5 no other changes? is this possible?

**Comment by dgrammatiko** [comment]
> can we split this pr please, one "simple" pr that only removes es5 no other changes?

The extra parts are the removal of the focus-visible or there are more? Can do

**Comment by HLeithner** [comment]
you removed chosen too, also I don't know the impact of changing template.js from defer:true to type:module

**Comment by dgrammatiko** [comment]
> I don't know the impact of changing template.js from defer:true to type:module

`type="module"` is also embedding `defer` but also it has narrower scope (a good thing). The template.js is only depends on `core.js` so it's completely safe.

> you removed chosen too

[chosen is dead](https://www.npmjs.com/package/chosen-js?activeTab=versions). It was dead back in 2015 when Joomla v4 was started so at some point this needs to be removed. That said I can revert this change (although I don't quite agree with the notion of distributing dead code)

Screenshot 2023-03-08 at 12 23 11

File: build/build-modules-js/settings.json (hunk 4 above)
Comment (dgrammatiko) [reply build/build-modules-js/settings.json]: Reverted

**Comment by HLeithner** [comment]
about chosen, we use it on our own site https://developer.joomla.org/joomlacode-archive/tracker-32.html with joomla 4 and the promise is that this should continue without issues. That brings me to 2 questions.

1. are templates using the es5 at the moment, you mentioned they are broken
2. is it possible to move chosen and focus-visible (if this is not replaced by something else automatically) to the b/c plugin?

**Comment by dgrammatiko** [comment]
> is it possible to move chosen and focus-visible (if this is not replaced by something else automatically) to the b/c plugin?

Both were reverted but I have no clue how can PHP influence some static assets

> are templates using the es5 at the moment, you mentioned they are broken

NO. All the methods (HTMLHelper, WAM) will only include assets (js,css) only if the files exist

**Comment by Fedik** [comment]
Looks good to me.
I would trash chosen, but if people still want it we can keep. Good that it not used by core anymore :)

**Comment by dgrammatiko** [comment]
> only final comment do we still not want to use rollup in most cases against the code so that we can use modern js syntax like async/await or similar if we want

Async/await is covered by the existing browserlist: https://caniuse.com/async-functions All versions are either lower. or the exact version (safari) for both modules and async/await

@wilsonge tbh I'm not sure if you're asking that rollup should do or NOT the bundling of every script. If you're asking for using it everywhere then it's already like that with only requirement that the source files end with `.es6.js` (the `.es5.js` could also handled by rollup, currently they are NOT, but that will require some changes (rollup will use the `iife` mode so the scripts should not be wrapped in an iife). If you're asking to opt out the `.es6.js` from rollup then that would break already a lot of scripts (everything that has an `import`) and cannot be done until [import maps](https://caniuse.com/import-maps) are universally supported (and still you need a polyfill, it exists). That said if the idea is to support import maps then there are changes that need to be done to the Web Asssets Manager so it could render the importmaps script, feature detect the US support, etc. Also keep in mind that unhandled scripts are slower due to their nature (the script will queue more fetches for all the imports...). 
Ok, that said the idea, since we decided to move to codebase to es6, was to at some point utilise importmaps. Maybe for Joomla 5 some groundwork could be done so J6 could roll with importmaps as the default option (someone needs to do this work)

**Comment by HLeithner** [comment]
OK I will merge this, can you write a explaination in the [Manual](https://github.com/joomla/Manual/tree/5.0) in [44-50 b/c break section](http://pr-28.manual.joomlacode.org/migrations/44-50/removed-backward-incompatibility) for people like me what the impact means for atum and for cassiopea?

Especially for extension developer? at the moment we load es5 and "es6" version (most of the time). Could there be some strange hacks by 3rd party which would be effected?

thanks

**Comment by dgrammatiko** [comment]
@HLeithner https://github.com/joomla/Manual/pull/85

> for people like me what the impact means for atum and for cassiopea?

There shouldn't be any change on the core templates

> Especially for extension developer? at the moment we load es5 and "es6" version (most of the time). Could there be some strange hacks by 3rd party which would be effected?

Practically with this PR we remove support for IE11 (or other UA that don't support ES2018). If anyone needs to add support for anything older than 2018 they need to ship their own scripts (and all the related dependencies)

**Comment by HLeithner** [comment]
seems the build script is not ready?

https://ci.joomla.org/joomla/joomla-cms/63089/2/2

[Error: ENOENT: no such file or directory, lstat '/******/src/build/tmp/1678611982/media/plg_multifactorauth_webauthn/js/webauthn-es5.min.js'] {

didn't investigated further yet

**Comment by HLeithner** [comment]
hmm the build folder is full of 'es5' references maybe they should all get removed?

**Comment by dgrammatiko** [comment]
> hmm the build folder is full of 'es5' references maybe they should all get removed?

Done

**Comment by dgrammatiko** [comment]
Screenshot 2023-03-12 at 12 36 50

I was expecting a bigger reduction on the distributed zip file but it is what it is

**Comment by HLeithner** [comment]
2mb compress javascript is really mich.

**Comment by HLeithner** [comment]
In the build script is also full of es5 stuff

**Comment by dgrammatiko** [comment]
> In the build script is also full of es5 stuff

Do you mean this:
Screenshot 2023-03-12 at 13 02 35

The build scripts still handle -es5.js as there are still some scripts that are not yet ES6, eg https://github.com/joomla/joomla-cms/blob/59d23bed3d8c61172bd755c24c3415bacd9d8c44/build/media_source/com_associations/js/sidebyside.es5.js, https://github.com/joomla/joomla-cms/tree/59d23bed3d8c61172bd755c24c3415bacd9d8c44/build/media_source/legacy/js, https://github.com/joomla/joomla-cms/tree/59d23bed3d8c61172bd755c24c3415bacd9d8c44/build/media_source/plg_editors_tinymce/js/plugins/dragdrop, etc

Therefore this script cannot be removed before someone puts some effort to do the conversion

**Comment by HLeithner** [comment]
OK thanks, we still have code jquery code^^

**Comment by dgrammatiko** [comment]
> OK thanks, we still have code jquery code^^

One file, one administrator view is not that bad...

**Comment by HLeithner** [comment]
Any chance you our maybe @Fedik has time to fix this?

the we may move jquery to the BC plugin. (Maybe a bad idea)

**Comment by dgrammatiko** [comment]
> then we may move jquery to the BC plugin. (Maybe a bad idea)

You mean registering the HTMLHelper::jquery through a legacy plugin?
What will be the benefit?

**Comment by HLeithner** [comment]
Getting rid of jquery ;) but yeah not much more

**Comment by HLeithner** [comment]
While developing the b/c plugin I noticed some deeper problem with the deprecation already done for JTable and HTMLHelper...

Moving htmlhelper for jquery to the plugin isn't hard, keeping the functionality is questionable, for example the HTMLHelper for jquery also has a `token` function which sets the csrf token for the jquery ajax call.

I would like to move this to a Webassetitem, but not sure if this this makes it better and how it have to be changed base on the Joomla own fetch function using the csrf token only for the current domain...

File: build/build-modules-js/settings.json [hunk 5]
```diff
@@ -630,39 +585,39 @@
...
+           {
+             "name": "chosen",
+             "type": "preset",
+             "dependencies": [
+               "chosen#style",
+               "chosen#script"
+             ]
+           }
+         ]
+       },
```
Comment (HLeithner) [inline build/build-modules-js/settings.json]: code style?

**Comment by brianteeman** [comment]
Not sure if its relevant but dont forget the debug plugin uses jquery

**Comment by dgrammatiko** [comment]
> Not sure if its relevant but dont forget the debug plugin uses jquery

Back in 2018 I was about to offer converting it to vanilla js but the author wasn't really interested: https://github.com/maximebf/php-debugbar/issues/368

> I would like to move this to a Webassetitem, but not sure if this this makes it better and how it have to be changed base on the Joomla own fetch function using the csrf token only for the current domain...

The CSRF on that jQuery script has the same exact issue (leaking) as the Joomla.response. But there are other issues with the Joomla.request that are also valid for $.ajax() (this is not a good place for this discussion).

BTW moving the jQuery to WAM is a B/C break (unless you have a dummy HTMLHelper::jquery that calls the WAM.

For anyone following the jQuery discussion I would like to mention that the maintainers have a plan for jQuery v4 with a bunch of B/C breaks all due to **security**:
- https://github.com/jquery/jquery/issues?q=is%3Aissue+is%3Aclosed+security
- https://github.com/jquery/jquery/issues?q=is%3Aissue+security+is%3Aopen
- https://github.com/jquery/jquery/milestone/7

File: build/build-modules-js/settings.json (hunk 5 above)
Comment (dgrammatiko) [reply build/build-modules-js/settings.json]: Fixed, maybe someone needs to redo https://github.com/joomla/joomla-cms/pull/39712

**Comment by HLeithner** [comment]
> Not sure if its relevant but dont forget the debug plugin uses jquery

not planning to remove jquery from joomla, only want to reduce HTMLHelpers because they are not the way to go (if possible).

but completely irrelevant to this pr at the moment. I plan to do multiple Pr for the b/c plugin, if you are interested you can look at my repo.

> BTW moving the jQuery to WAM is a B/C break (unless you have a dummy HTMLHelper::jquery that calls the WAM.

HTMLHelper::jquery already already calls the WAM, so no b/c on this, and as sad I would move the jquery htmlhelper to the plugin (not removing it in 5.0). will be an one pr.

> regarding jquery 4

if it comes in time I would add it to joomla 5.0 not sure in which form yet.

**Comment by Fedik** [comment]
In case when people worries about b.c. break due to removed assets. We can keep the asset definations, but empty, with flag "deprecated". And remove all es5 js files.

Example for bootstrap,
Before:
```
{
  "name": "bootstrap.es5",
  "type": "script",
  "uri": "bootstrap-es5.min.js",
  "dependencies": [
    "core"
  ],
  "attributes": {
    "nomodule": true,
    "defer": true
  }
},
```
After:
```
{
  "name": "bootstrap.es5",
  "type": "script",
  "deprecated": true,
  "uri": "",
  "dependencies": [
    "core"
  ],
},
```
This dummies will save the site when some template use it in unknown reasons.

**Comment by dgrammatiko** [comment]
>  We can keep the asset definations, but empty

I was thinking about that as well, so maybe I should update the PR to eliminate any B/C considerations

**Comment by HLeithner** [comment]
I tested this with one of my customer sites with my own template using joomla bootstrap and works as expected, I merge this for now and hope for more real world testing for alpha2
